bootstrap/kubeadm/api/v1alpha3/conversion.go

```diff
@@ -74,3 +74,16 @@ func (dst *KubeadmConfigTemplateList) ConvertFrom(srcRaw conversion.Hub) error {
 func Convert_v1alpha3_KubeadmConfigStatus_To_v1alpha4_KubeadmConfigStatus(in *KubeadmConfigStatus, out *kubeadmbootstrapv1alpha4.KubeadmConfigStatus, s apiconversion.Scope) error { //nolint
 	return autoConvert_v1alpha3_KubeadmConfigStatus_To_v1alpha4_KubeadmConfigStatus(in, out, s)
 }
+
+// Convert_v1alpha4_KubeadmConfigSpec_To_v1alpha3_KubeadmConfigSpec is an autogenerated conversion function.
+func Convert_v1alpha4_KubeadmConfigSpec_To_v1alpha3_KubeadmConfigSpec(in *kubeadmbootstrapv1alpha4.KubeadmConfigSpec, out *KubeadmConfigSpec, s apiconversion.Scope) error { //nolint
+	if err := autoConvert_v1alpha4_KubeadmConfigSpec_To_v1alpha3_KubeadmConfigSpec(in, out, s); err != nil {
+		return err
+	}
+
+	// To make fuzzy tests pass. Conversion from v1alpha4 to v1alpha3 is actually lossy, as IgnitionConfig field
+	// has been added in v1alpha4.
+	in.IgnitionConfig = nil
```

Comment:
Yeah, I guess ideally it shouldn't be lossy, but IIRC I shouldn't change v1alpha3 types at this point.

If data can be stored in annotation during conversion, I guess this will be fine, I'll have a look. Thanks for the pointer!

I only made it lossy so the unit tests are passing, I couldn't find a better way :smile:

Reply:
@JoelSpeed I've tried this approach at the very beginning actually and tests are still failing with the following error:
```
$ go test ./bootstrap/kubeadm/api/v1alpha3/
--- FAIL: TestFuzzyConversion (0.00s)
    --- FAIL: TestFuzzyConversion/for_KubeadmConfig (0.00s)
        --- FAIL: TestFuzzyConversion/for_KubeadmConfig/hub-spoke-hub (0.00s)
            conversion.go:193:
                  &v1alpha4.KubeadmConfig{
                        TypeMeta:   {},
                        ObjectMeta: {Name: "饿Ă|埍ǪġȤ÷莅ư\u00a0İƧŀ浩Ȳ郆", Namespace: "曐8ʪɕ]I", SelfLink: "劗诺@lŐɏȈĚZs3臍适ÇŎ-ɂ", UID: "5弒{B审蓖B=_襳ȣwǩí鷴5俽", ...},
                        Spec: v1alpha4.KubeadmConfigSpec{
                                ... // 11 identical fields
                                Verbosity:                &1324947815,
                                UseExperimentalRetryJoin: true,
                -               IgnitionConfig: &v1alpha4.IgnitionConfig{
                -                       ContainerLinuxConfig: &v1alpha4.ContainerLinuxConfig{AdditionalConfig: "澿ÝŦ氭NƬĹA慰U"},
                -               },
                +               IgnitionConfig: nil,
                        },
                        Status: {Ready: true, FailureReason: "#ȦK", FailureMessage: "牊ùÀyȎǾ殴倴跴ƇduÅy濬´I_Ù", ObservedGeneration: -5299259527439887890, ...},
                  }

                Expected
                    <bool>: false
                to be true
    --- FAIL: TestFuzzyConversion/for_KubeadmConfigTemplate (0.00s)
        --- FAIL: TestFuzzyConversion/for_KubeadmConfigTemplate/hub-spoke-hub (0.00s)
            conversion.go:193:
                  &v1alpha4.KubeadmConfigTemplate{
                        TypeMeta:   {},
                        ObjectMeta: {Name: "ɨåAF]č鉘曀", GenerateName: "駮qÎ歇=Ș岐鲑訫fª", Namespace: `ǼL<嶇ʎ\`, SelfLink: "+1Ï岤C4'©堢}ɩ檉ɨ", ...},
                        Spec: v1alpha4.KubeadmConfigTemplateSpec{
                                Template: v1alpha4.KubeadmConfigTemplateResource{
                                        Spec: v1alpha4.KubeadmConfigSpec{
                                                ... // 11 identical fields
                                                Verbosity:                nil,
                                                UseExperimentalRetryJoin: false,
                -                               IgnitionConfig: &v1alpha4.IgnitionConfig{
                -                                       ContainerLinuxConfig: &v1alpha4.ContainerLinuxConfig{AdditionalConfig: "IÑF亢絃梲ŻÆL"},
                -                               },
                +                               IgnitionConfig: nil,
                                        },
                                },
                        },
                  }

                Expected
                    <bool>: false
                to be true
FAIL
FAIL    sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3  0.016s
FAIL
```

Reply:
I had a look at this and came up with a branch that does have this working, https://github.com/JoelSpeed/cluster-api/commits/ignition-support

I've added two separate commits here, the [first](https://github.com/JoelSpeed/cluster-api/commit/19e2b77f74ff6595561e311611507a54c6d153e3) marshals the data on the down conversion (looking at the above result, I think you may have missed this?) and then restores the data on the up conversion.

The [second](https://github.com/JoelSpeed/cluster-api/commit/0a398274aedfcac5f697e8ef56d05fa100c2f263) fixes some fuzzing issue. Because storing the data in this way marshals and unmarshals JSON, the fuzzed data has to be compatible with the objects json marshalling. In this particular case, the `kubeadmv1beta1.BootstrapTokenString` has an issue because it expects a particular format (`[a-z0-9]{6}\.[a-z0-9]{16}`), so we have to manually create a fuzzer for this to ensure that it can be marshalled and unmarshalled without error.

This extra fuzzing is the first to be added/needed for the v1alpha3/v1alpha4 conversion so I'd suggest we get a node from @vincepri on this topic before picking those changes to this PR

Reply:
Thanks for looking into this @JoelSpeed, highly appreciated :) I was missing lines:
```
dst.Spec.IgnitionConfig = restored.Spec.IgnitionConfig
```
and
```
dst.Spec.Template.Spec.IgnitionConfig = restored.Spec.Template.Spec.IgnitionConfig
```

in my implementation :)

If I understand correctly, I'll wait for feedback from @vincepri before adapting changes you suggested to this PR.

Reply:
We should restore the field when up-converting, rather than removing it when down-converting. 

For example: 
- Stores the new version in an annotation https://github.com/kubernetes-sigs/cluster-api/blob/release-0.3/bootstrap/kubeadm/api/v1alpha2/conversion.go#L81-L84
- https://github.com/kubernetes-sigs/cluster-api/blob/release-0.3/bootstrap/kubeadm/api/v1alpha2/conversion.go#L33-L39 Restores the fields that are new.

Reply:
@vincepri sure, this is what @JoelSpeed suggested. But what do you think about https://github.com/JoelSpeed/cluster-api/commit/0a398274aedfcac5f697e8ef56d05fa100c2f263 ? Should I also pull it as part of this PR then?

Reply:
If you don't pull in  JoelSpeed@0a39827, you'll need an equivalent fuzzer otherwise you'll get JSON marshalling errors, perhaps one already exists though that I couldn't find when I was trying this out

Reply:
I rebased and included your patch @JoelSpeed, so code now retains the data.

Reply:
I'll mark it as resolved if there is no further comments.